examples/modular-drivers/src/bme280_ulp.rs

```diff
@@ -0,0 +1,970 @@
+use std::{
+    cell::UnsafeCell,
+    ffi::c_void,
+    ops::Shl,
+    ptr::addr_of,
+    sync::{Arc, Mutex},
+    thread::sleep,
+    time::Duration,
+};
+
+use micro_rdk::google::protobuf::{value, Value};
+use micro_rdk::{
+    common::{
+        config::{AttributeError, ConfigType, Kind},
+        i2c::I2cHandleType,
+        registry::{self, ComponentRegistry, Dependency, RegistryError},
+        sensor::{GenericReadingsResult, Readings, Sensor, SensorError, SensorType},
+    },
+    esp32::esp_idf_svc::{hal::ulp::UlpDriver, sys::EspError},
+    DoCommand,
+};
+use thiserror::Error;
+
+// contains the address of global variables defined in the ULP program
+include!(concat!(env!("OUT_DIR"), "/ulp.rs"));
+
+static ULP_PROGRAM: &[u8] = include_bytes!(concat!(env!("OUT_DIR"), "/ulp.bin"));
+
+struct UnsafeRtcMemory<T>(UnsafeCell<T>);
+unsafe impl<T> Sync for UnsafeRtcMemory<T> {}
+
+static SAMPLE_ARRAY_SIZE: usize = 32;
+#[unsafe(link_section = ".rtc.force_slow")]
+static SAMPLE_ARRAY: UnsafeRtcMemory<[u32; SAMPLE_ARRAY_SIZE]> =
+    UnsafeRtcMemory(UnsafeCell::new([0xFFFFFFFF_u32; SAMPLE_ARRAY_SIZE]));
+
+#[derive(Error, Debug)]
+pub enum BME280Error {
+    #[error("pin {0} is not an RTC pin")]
+    BME280UlpInvalidRTCPin(i32),
+    #[error("pin {0} is not valid pin")]
+    BME280UlpInvalidPin(i32),
+    #[error("error configuring driver {0}")]
+    BME280I2cDriverError(EspError),
+    #[error("cannot instantiate UlpDriver: {0}")]
+    BME280UlpDriverError(EspError),
+    #[error("error loading Ulp Program: {0}")]
+    BME280UlpLoadError(EspError),
+    #[error("unable to configure rtc pin {0} cause {1}")]
+    BME280UlpRtcPinConfigError(i32, EspError),
+    #[error("cannot start ulp program {0}")]
+    BME280UlpCannotStart(EspError),
+    #[error("invalid config value")]
+    BME280InvalidConfigValue,
+}
+
+struct ULPResultsMemory<'a> {
+    memory: &'a [u32],
+    offset: usize,
+}
+
+impl<'a> From<&'a UnsafeRtcMemory<[u32; SAMPLE_ARRAY_SIZE]>> for ULPResultsMemory<'a> {
+    fn from(value: &'a UnsafeRtcMemory<[u32; SAMPLE_ARRAY_SIZE]>) -> Self {
+        let ptr = unsafe { &*value.0.get() };
+        Self {
+            memory: ptr,
+            offset: 0,
+        }
+    }
+}
+
+impl Iterator for ULPResultsMemory<'_> {
```

Comment:
Is this the iterator that the data sync task is supposed to read from?

Reply:
no we would use it to construct Readings with their associated timestamp